src/main/scala/viper/gobra/frontend/InformativeErrorListener.scala

```diff
@@ -0,0 +1,324 @@
+import viper.gobra.frontend.Source.FromFileSource
+import viper.gobra.reporting.ParserError
+import viper.silver.ast.SourcePosition
+
+import java.nio.file.Path
+import scala.collection.mutable.ListBuffer
+
+class InformativeErrorListener(val messages: ListBuffer[ParserError], val source: Source) extends BaseErrorListener {
+
+  /**
+    *
+    * @param recognizer The recognizer that encountered the error
+    * @param offendingSymbol The symbol that caused the error
+    * @param line The line number in the Source
+    * @param charPositionInLine The column in the line
+    * @param msg The message emitted by the [[org.antlr.v4.runtime.ANTLRErrorListener]]
+    * @param e The specific [[RecognitionException]] thrown by the parser
+    */
+  override def syntaxError(recognizer: Recognizer[_, _], offendingSymbol: Any, line: Int, charPositionInLine: Int, msg: String, e: RecognitionException): Unit = {
+    // We don't analyze Lexer errors any further: The defaults are sufficient
+    val error = recognizer match {
+      case lexer: Lexer => DefaultErrorType()(LexerErrorContext(lexer, null, line, charPositionInLine, msg))
+      case parser: Parser => {
+        analyzeParserError(ParserErrorContext(parser, offendingSymbol.asInstanceOf[Token], line, charPositionInLine, msg), e)
+      }
+    }
+
+    // Depending on the source, get the applicable type of position information
+    val pos = source match {
+      case source: FileSource => Some(SourcePosition(Path.of(source.name), line, charPositionInLine))
+      case source: FromFileSource => Some(SourcePosition(source.path, line, charPositionInLine))
+      case _ => None
+    }
+
+    // Wrap the error in Gobra's error class
+    val message = Some(ParserError(error.full, pos))
+    messages.appendAll(message)
+  }
+
+  /**
+    * Check the given error context against a set of known patterns to check for specific common errors.
+    * @param context
+    * @param e
+    * @return
+    */
+  def analyzeParserError(implicit context : ParserErrorContext, e : RecognitionException): ErrorType = {
+    e match {
+      case exception: FailedPredicateException => analyzeFailedPredicate(context, exception)
+      case exception: InputMismatchException => analyzeInputMismatch(context, exception)
+      case exception: NoViableAltException => analyzeNoViable(context, exception)
+      case null => context.msg match {
+        case extraneous() => analyzeExtraneous(context)
+        case missing() => DefaultErrorType()
+      }
+      case _ => DefaultErrorType()
+    }
+  }
+
+
+  /**
+    * Failed predicate errors are emitted when the no viable alternatives remain, as all conflict with the input
+    * or contain a semantic predicate that evaluated to false. Because the only predicate present in Gobra's grammar
+    * is responsible for inducing semicolons, we know that we are at a point where a statement could have ended, but
+    * another token was discovered.
```

Comment:
```suggestion
    * Failed predicate errors are emitted when no viable alternatives remain, as all alternatives conflict with the input
    * or contain a semantic predicate that evaluated to false. Because the only predicate present in Gobra's grammar
    * is responsible for inducing semicolons, we know that if a predicate failed, we are at a point where a statement could have ended, but
    * another token was discovered.
```